aiida/schedulers/scheduler.py

```diff
@@ -232,7 +232,13 @@ def _get_run_line(self, codes_info, codes_run_mode):
             else:
                 stderr_str = f'2> {escape_for_bash(code_info.stderr_name)}' if code_info.stderr_name else ''
 
-            output_string = f'{command_to_exec} {stdin_str} {stdout_str} {stderr_str}'
+            command_string = f'{command_to_exec} {stdin_str} {stdout_str} {stderr_str}'
+
+            if code_info.folder_path:
+                run_in_folder = f'cd {escape_for_bash(code_info.folder_path)}'
+                output_string = f'({run_in_folder}; {command_string})'
```

Comment:
I think this might also need to run a `cd -` (i.e. to change to the base directory again. Best would not even to use `cd -` but to use the explicit path which would be more robust. Not sure how that is defined in this part of the scheduler though) because the `append_text` may include code that depends on being run in the root dir of the job.

Reply:
The ``cd -`` should not be needed in principle, since running using ``(cd folder; code.x ...)`` will run basically the code in ``folder`` from the root one in a sub shell. But interesting point regarding ``append_text``. Don't know if it can cause problems.

Reply:
Responded to the general point in the other comment, but it would be good to test a bit with the `append_text`. It seems like it just executes a subshell and then exits to the previous state, so it should be good I think.

Reply:
Yes, if executed in a subshell, in the next line it's back to the original folder so I don't see a problem here.

Reply:
A thing that makes me wondering is: using ``cd {escape_for_bash(code_info.folder_path)}``, one can potentially create and run in any path. Shall we limit it to a subfolder of the main folder? E.g. ``cd ./{escape_for_bash(code_info.folder_path)}``?

Reply:
Yes, we should validate that the value is a relative path or else we except. This is why I suggested including `relative` in the option name so it is absolutely clear.

Reply:
Where can we put the control in your opinion?  Because the root folder info is in CalcJob, isn't it?

Reply:
I agree that adding a check is good. Note that adding `./` in front is not enough - e.g. if the folder_path is `../somewhere_else`. Maybe a good place to put this, if we agree this should always be done, is in `presubmit`? https://github.com/aiidateam/aiida-core/blob/fe1acf9f6d425d1be973e3ecd33f19760f59e77c/aiida/engine/processes/calcjobs/calcjob.py#L573 

One way to check this can be to get the abs path and then check if the common prefix is the same, as we do here: https://github.com/aiidateam/aiida-core/blob/fe1acf9f6d425d1be973e3ecd33f19760f59e77c/aiida/common/folders.py#L49-L60 even if I think that recent python 3.x has even a method to check if a path is a subpath? if you decide you always prepend `./` maybe checking that there is no path component being .. might be ok, even if I'm not sure it's enough

Reply:
We should really just the [`pathlib`](https://docs.python.org/3/library/pathlib.html) module. We should simply require that the `relative_run_directory` is a `pathlib.Path` object and then we can use the following check
```python
assert (workdir / calc_info.relative_run_directory).is_relative_to(workdir)
```
This should catch all the possible cases. Quick demo (not sure if I caught all possibilities, but you get the picture):
```python
In [1]: from pathlib import Path

In [2]: workdir = Path('/tmp/scratch/deadbeef')

In [3]: for relative_path in [Path('relative'), Path('./relative'), Path('../deadbeef/relative'), Path('../../ilegal'), Path('/tmp/absolute')]:
   ...:     path = workdir / relative_path
   ...:     print(path.is_relative_to(workdir))
   ...: 
True
True
True
True
False
```

After the check, we just make sure that we pass the resolved run directory to the scheduler plugin. The problem here is that the scheduler plugin is responsible for respecting this feature but we cannot enforce this since the plugins can be defined out side of `aiida-core`. This of course is subideal but we have seen this problem before and there isn't a good solution for now.

Reply:
I have 2 comments/issues with this approach.

1. `is_relative_to` exists [only since 3.9](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.is_relative_to) so I think we cannot use it as we support also older python versions.

2. One or your examples does not convince me as doing what we want:
   ```python
   from pathlib import Path
   workdir = Path('/tmp/scratch/deadbeef')
   path = workdir / Path('../../ilegal')
   print(path.is_relative_to(workdir))
```
This returns true. But I think the check we want should return False, no? I think it just checks the string.
Indeed, if we first resolve it, then it returns False: `print(path.resolve().is_relative_to(workdir))`.

So probably we should do it.

Finally - I just wanted to mention that this assumes that paths on the supercomputer are managed in the same way as in the computer where AiiDA is running. I think we make the same assumption elsewhere (and probably it's OK if people run in Windows inside the WSL?) so maybe OK also to do it here. Maybe in the future we should also check/ask with OS is on the remote host and use WindowsPath/PosixPath correctly. But I guess this is for the future, when someone will raise the issue...

Reply:
>     `is_relative_to` exists [only since 3.9](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.is_relative_to) so I think we cannot use it as we support also older python versions.

Damn, missed that. Alternatively we could use [`relative_to`](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.relative_to) which is the same but raises instead of returning `False`.

>     One or your examples does not convince me as doing what we want:

You are right of course, I intended that to return `False` but it doesn't and missed it. Resolving would work, but the downside (and the reason I didn't use it initially) is that this is a method of `Path` and not `PurePath`, so that operation actually needs access to the filesystem and is not a purely lexical operation. This is not possible in this case since we are operating on paths on a remote file system. So maybe my approach indeed won't work. But I would still try to find a solution using `pathlib` instead of `os.